tools/sof-perf-analyzer.py

```diff
@@ -0,0 +1,253 @@
+        if match_obj := re.search(r'\[.+\] <(dbg|inf|wrn|err)>', line):
+            # Sometimes, A trace output may be incomplete and mixed with next
+            # trace output in a single line, for example: '[    0.071590] <inf>
+            # pipe[    0.071751] <inf> host_comp: host_get_copy_bytes_normal: comp:1
+            # 0x40003 no bytes to copy, available samples: 0, free_samples: 384'.
+            # Under this circumstance, we ignore the first trace and extract the
+            # second trace from the line. This is done by regarding the end position
+            # of the matched substring as sentinel, timestamp and trace level are
+            # both some specific offset from the sentinel.
+            span_end_pos = match_obj.span()[1]
+            trace_lvl = line[span_end_pos - 4: span_end_pos - 1]
+            timestamp = float(line[span_end_pos - 19: span_end_pos - 7].strip())
+
+            # The rest after removing timestamp and log level
+            rest = line[span_end_pos + 1:].split(': ')
+            ctx, func = rest[0:2]
+            msg = ': '.join(rest[2:]).strip()
+            yield TraceItem(timestamp, trace_lvl, ctx, func, msg)
+
+def process_trace_file():
+    '''The top-level caller for processing the trace file'''
+    with open(args.filename, 'r', encoding='utf8') as file:
+        trace_item_gen = make_trace_item(file)
+        trace_prev = get_first_valid_trace(trace_item_gen)
+        if trace_prev is None:
+            raise Exception('No valid trace in provided file')
+        for trace_curr in trace_item_gen:
+            # pylint: disable=W0603
+            global ts_shift
+            old_ts_shift = ts_shift
+            # On wrap happened, the timestamp of current trace should be much more smaller
+            # than the previous one. In practice, it is possible that the timestamp of
+            # current trace is slightly smaller than the previous one, this could be a
+            # bug in SOF. Add a 50s shift to make sure timestamp correction work properly.
+            if trace_curr.timestamp < trace_prev.timestamp - 50:
+                ts_shift = ts_shift + UINT32_MAX / DSP_TIMER
+            trace_prev.timestamp += old_ts_shift
+            dispatch_trace_item(trace_prev)
+            trace_prev = trace_curr
+        trace_prev.timestamp += ts_shift
+        dispatch_trace_item(trace_prev)
+
+def process_dmesg_file():
+    '''Process the dmesg to get the component ID to component name mapping,
+    component name is acquired from the line that contains 'Create widget',
+    component ID is acquired from the next line. Example:
+    [  334.818435] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    Create widget host-copier.0.playback instance 0 - pipe 1 - core 0
+    [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]'''
```

Comment:
What happens when the input given to this function is made of multiple FW runs with overlapping and possibly conflicting information? Fails or last one wins? This needs to be at least documented.